File: backend/schemas/calculators.py
from pydantic import BaseModel
from typing import List


# ── SIP ──────────────────────────────────────────────────────────────────────

class SipInput(BaseModel):
    monthlyInvestment: float
    annualReturnRate: float
    tenureYears: int


class SipResult(BaseModel):
    totalInvested: float
    estimatedReturns: float
    maturityValue: float


# ── EMI ──────────────────────────────────────────────────────────────────────

class EmiInput(BaseModel):
    principalAmount: float
    annualInterestRate: float
    tenureMonths: int


class AmortizationRow(BaseModel):
    month: int
    principal: float
    interest: float
    balance: float


class EmiResult(BaseModel):
    emi: float
    totalAmount: float
    totalInterest: float
    amortizationSchedule: List[AmortizationRow]


# ── Retirement ────────────────────────────────────────────────────────────────

class RetirementInput(BaseModel):
    currentAge: int
    retirementAge: int
    monthlyExpenses: float
    currentSavings: float
    expectedReturnRate: float
    inflationRate: float


class RetirementResult(BaseModel):
    corpusRequired: float
    monthlySavingsNeeded: float
    projectedCorpus: float
    yearsToRetirement: int


# ── CTC In-hand ───────────────────────────────────────────────────────────────

class CtcInhandInput(BaseModel):
    ctcAnnual: float
    pfContribution: float
    professionalTax: float
    otherDeductions: float


class BreakdownRow(BaseModel):
    label: str
    amount: float


class CtcInhandResult(BaseModel):
    grossMonthly: float
    deductions: float
    netMonthly: float
    annualTakeHome: float
    breakdown: List[BreakdownRow]


# ── Tax Regime ────────────────────────────────────────────────────────────────

class TaxRegimeInput(BaseModel):
    annualIncome: float
    hraReceived: float
    rentPaid: float
    section80c: float
    section80d: float
    otherDeductions: float


class SlabRow(BaseModel):
    slab: str
    tax: float


class TaxRegimeResult(BaseModel):
    oldRegimeTax: float
    newRegimeTax: float
    recommendation: str
    savings: float
    oldRegimeBreakdown: List[SlabRow]
    newRegimeBreakdown: List[SlabRow]

File: src/api/calculators.ts
import apiClient from './client';

// ── SIP ────────────────────────────────────────────────────────

export interface SipInput {
  monthlyInvestment: number;
  annualReturnRate: number;
  tenureYears: number;
}

export interface SipResult {
  totalInvested: number;
  estimatedReturns: number;
  maturityValue: number;
}

export async function sip(input: SipInput): Promise<SipResult> {
  const { data } = await apiClient.post<SipResult>('/calculators/sip', input);
  return data;
}

// ── EMI ────────────────────────────────────────────────────────

export interface EmiInput {
  principalAmount: number;
  annualInterestRate: number;
  tenureMonths: number;
}

export interface EmiResult {
  emi: number;
  totalAmount: number;
  totalInterest: number;
  amortizationSchedule: Array<{
    month: number;
    principal: number;
    interest: number;
    balance: number;
  }>;
}

export async function emi(input: EmiInput): Promise<EmiResult> {
  const { data } = await apiClient.post<EmiResult>('/calculators/emi', input);
  return data;
}

// ── Retirement ─────────────────────────────────────────────────

export interface RetirementInput {
  currentAge: number;
  retirementAge: number;
  monthlyExpenses: number;
  currentSavings: number;
  expectedReturnRate: number;
  inflationRate: number;
}

export interface RetirementResult {
  corpusRequired: number;
  monthlySavingsNeeded: number;
  projectedCorpus: number;
  yearsToRetirement: number;
}

export async function retirement(input: RetirementInput): Promise<RetirementResult> {
  const { data } = await apiClient.post<RetirementResult>('/calculators/retirement', input);
  return data;
}

// ── CTC In-hand ────────────────────────────────────────────────

export interface CtcInhandInput {
  ctcAnnual: number;
  pfContribution: number;
  professionalTax: number;
  otherDeductions: number;
}

export interface CtcInhandResult {
  grossMonthly: number;
  deductions: number;
  netMonthly: number;
  annualTakeHome: number;
  breakdown: Array<{ label: string; amount: number }>;
}

export async function ctcInhand(input: CtcInhandInput): Promise<CtcInhandResult> {
  const { data } = await apiClient.post<CtcInhandResult>('/calculators/ctc-inhand', input);
  return data;
}

// ── Tax Regime ─────────────────────────────────────────────────

export interface TaxRegimeInput {
  annualIncome: number;
  hraReceived: number;
  rentPaid: number;
  section80c: number;
  section80d: number;
  otherDeductions: number;
}

export interface TaxRegimeResult {
  oldRegimeTax: number;
  newRegimeTax: number;
  recommendation: 'old' | 'new';
  savings: number;
  oldRegimeBreakdown: Array<{ slab: string; tax: number }>;
  newRegimeBreakdown: Array<{ slab: string; tax: number }>;
}

export async function taxRegime(input: TaxRegimeInput): Promise<TaxRegimeResult> {
  const { data } = await apiClient.post<TaxRegimeResult>('/calculators/tax-regime', input);
  return data;
}
